[CARBONDATA-1360]  fix for update table failures for struct data type

ashwini-krishnakumar opened the pull request:
Problem: updates not working for complex data type struct. escape characters used for internal delimiters was causing the update failures and null updates .
Fix : escape characters used for internal delimiters was causing the update failures and null updates to fix this escape characters are replaced before update processing.

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[CARBONDATA-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).
 - [ ] Testing done
 
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - What manual testing you have done?
        - Any additional information to help reviewers in testing this change.
         
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
                 
---

CarbonDataQA commented:
Can one of the admins verify this patch?

ravipesala commented:
Can one of the admins verify this patch?

CarbonDataQA commented:
Can one of the admins verify this patch?

jackylk commented:
retest this please

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/511/

ravipesala commented:
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1140/

ashwini-krishnakumar commented:
Problem: updates not working for complex data type .
    Fix : escaspe characters used for internal demiliters was causing the update failures and null updates to fix this escaspe characters are replaced before update processing.
    Changes :1. added testcase to test update on complex struct data type with two level delimiters
             2. fixed update failures for array and struct datatype
             3. added testcase for update with array data type

ravipesala commented:
Can one of the admins verify this patch?

ashwini-krishnakumar commented:
pleas note - This PR has a dependency of PR 1101 to be merged first .

ravipesala commented:
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1286/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1420/

ashwini-krishnakumar commented:
retest this please

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2433/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3506/

ashwini-krishnakumar commented:
retest this please

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3715/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2475/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3517/

zzcclp commented:
@ashwini-krishnakumar I run your test cases with my pr and all are passed too.

zzcclp commented:
@ashwini-krishnakumar [PR#1962](https://github.com/apache/carbondata/pull/1962) has been merged, I think it don't need to change  CarbonScalaUtil.scala and just remain the test case.

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/3067/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4311/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5199/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/3989/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7136/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5912/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/63/

CarbonDataQA commented:
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8812/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/745/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/567/

CarbonDataQA commented:
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8866/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2015/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2192/

CarbonDataQA commented:
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10267/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/90/

CarbonDataQA commented:
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/89/

CarbonDataQA1 commented:
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/87/

CarbonDataQA2 commented:
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5206/
